### PR TITLE
Added assignment 0

### PR DESCRIPTION
**Pull request by shahpratikr**

### CODE REVIEW

File: assignment-0/controller.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+type controller struct {
+	Client   *kubernetes.Clientset
+	Informer cache.SharedIndexInformer
+}
+
+func newController(client *kubernetes.Clientset, informerFactory informers.SharedInformerFactory) *controller {
+	deploymentInformer := informerFactory.Apps().V1().Deployments().Informer()
+	deploymentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			createService(client, obj)
```
Comment (viveksinghggits) [inline assignment-0/controller.go:26]: If one call of `createService` takes 10 seconds. And we create some more deployments within those 10 seconds, what would be the behaviour of the code?

File: assignment-0/controller.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func (c *controller) worker() {
+	for c.processItem() {
+	}
+}
+
+func (c *controller) processItem() bool {
+	return true
+}
```
Comment (viveksinghggits) [inline assignment-0/controller.go:57]: Do we really need this if we are just returning `true` always.

File: assignment-0/controller.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+				Labels:    deployment.Labels,
+			},
+			Spec: corev1.ServiceSpec{
+				Ports:    servicePorts,
+				Selector: deployment.Labels,
+				Type:     "LoadBalancer",
+			},
+		}, metav1.CreateOptions{})
+	if err != nil {
+		panic(err.Error())
```
Comment (viveksinghggits) [inline assignment-0/controller.go:95]: Using panic directly is discouraged we should return/handle the error instead.

File: assignment-0/controller.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      deployment.Name + "-service",
+				Namespace: deployment.Namespace,
+				Labels:    deployment.Labels,
+			},
+			Spec: corev1.ServiceSpec{
+				Ports:    servicePorts,
+				Selector: deployment.Labels,
+				Type:     "LoadBalancer",
+			},
```
Comment (viveksinghggits) [inline assignment-0/controller.go:92]: This service object should be created separately and maybe in a new helper function to make code readable and that would help to write  unit tests as well.

File: assignment-0/controller.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		}, metav1.CreateOptions{})
+	if err != nil {
+		panic(err.Error())
+	}
+}
+
+func deleteService(clientset *kubernetes.Clientset, obj interface{}) {
+	deployment := obj.(*appsv1.Deployment)
+	// check if service is present, if it is then only delete
+	_, err := clientset.CoreV1().Services(deployment.Namespace).Get(
```
Comment (viveksinghggits) [inline assignment-0/controller.go:102]: In this case the entire codebase is in single file, so things are easier but if we had multiple files and we changed how we generate the service name, how would we know that we have to change this place as well. That is the reason we should have a common function to generate the service name and that can be used at both the places.
These things don't seem important but they are, very.

File: assignment-0/main.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func main() {
+	var (
+		config *rest.Config
+		err    error
+	)
+	kubeconfig := flag.String("kubeconfig", "/home/pratik/.kube/config",
+		"location for kubeconfig file")
```
Comment (viveksinghggits) [inline assignment-0/main.go:21]: `/home/pratik` would not be there on my machine.

File: assignment-0/controller.go (hunk 1 above)
Comment (shahpratikr) [reply assignment-0/controller.go:26]: added a queue to address this

File: assignment-0/controller.go (hunk 2 above)
Comment (shahpratikr) [reply assignment-0/controller.go:57]: modified this function

File: assignment-0/controller.go (hunk 3 above)
Comment (shahpratikr) [reply assignment-0/controller.go:95]: fixed

File: assignment-0/controller.go (hunk 4 above)
Comment (shahpratikr) [reply assignment-0/controller.go:92]: fixed

File: assignment-0/controller.go (hunk 5 above)
Comment (shahpratikr) [reply assignment-0/controller.go:102]: fixed

File: assignment-0/main.go (hunk 6 above)
Comment (shahpratikr) [reply assignment-0/main.go:21]: fixed

File: assignment-0/pkg/controller/controller.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type Controller struct {
+	Client        *kubernetes.Clientset
+	Informer      cache.SharedIndexInformer
+	Queue         workqueue.RateLimitingInterface
+	ServiceDelete bool
```
Comment (viveksinghggits) [inline assignment-0/pkg/controller/controller.go:22]: can you add comment on what exactly `ServiceDelete` does?

Comment (shahpratikr) [reply assignment-0/pkg/controller/controller.go:22]: added

File: assignment-0/pkg/controller/controller.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+				Protocol: containerPort.Protocol,
+				Port:     containerPort.ContainerPort,
+			}
+			servicePorts = append(servicePorts, port)
+		}
+	}
+	return servicePorts
+}
+
+func (c *Controller) createServiceInKubernetes(deployment *appsv1.Deployment) error {
```
Comment (viveksinghggits) [inline assignment-0/pkg/controller/controller.go:119]: this should be just `createService` I think. It's implicit that we are going to create the service in Kubernetes.

File: assignment-0/pkg/controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+	<-ch
+}
+
+func (c *Controller) worker() {
+	for c.processItems() {
+	}
+}
+
+func (c *Controller) getServiceName(deployment *appsv1.Deployment) string {
```
[the remark this answers is not in the document]
Comment (shahpratikr) [reply assignment-0/pkg/controller/controller.go:54]: @viveksinghggits didn't get you

File: assignment-0/pkg/controller/controller.go (hunk 8 above)
Comment (shahpratikr) [reply assignment-0/pkg/controller/controller.go:110]: I have another function with same name. hence used `createServiceInKubernetes` as a name

**Pull request merged**